server/apidata/apidata.go

```diff
@@ -0,0 +1,247 @@
+	BinSizes = []string{"24h", "1h", "5m"}
+	// Our internal millisecond representation of the bin sizes.
+	binSizes []uint64
+	bin5min  uint64 = 60 * 5 * 1000
+	started  uint32
+)
+
+// DBSource is a source of persistent data. DBSource is used to prime the
+// caches at startup.
+type DBSource interface {
+	LoadEpochStats(base, quote uint32, caches []*db.CandleCache) error
+}
+
+// MarketSource is a source of market information. Markets are added after
+// construction but before use using the AddMarketSource method.
+type MarketSource interface {
+	EpochDuration() uint64
+	Base() uint32
+	Quote() uint32
+}
+
+// BookSource is a source of order book information. The BookSource is added
+// after construction but before use.
+type BookSource interface {
+	Book(mktName string) (*msgjson.OrderBook, error)
+}
+
+// DataAPI is a data API backend.
+type DataAPI struct {
+	db             DBSource
+	epochDurations map[string]uint64
+	bookSource     BookSource
+
+	spotsMtx sync.RWMutex
+	spots    map[string]json.RawMessage
+
+	cacheMtx     sync.RWMutex
+	marketCaches map[string]map[uint64]*db.CandleCache
+	cache5min    *db.CandleCache
+}
+
+// NewDataAPI is the constructor for a new DataAPI.
+func NewDataAPI(ctx context.Context, dbSrc DBSource) *DataAPI {
+	s := &DataAPI{
+		db:             dbSrc,
+		epochDurations: make(map[string]uint64),
+		spots:          make(map[string]json.RawMessage),
+		marketCaches:   make(map[string]map[uint64]*db.CandleCache),
+	}
+
+	if atomic.CompareAndSwapUint32(&started, 0, 1) {
+		comms.RegisterHTTP(msgjson.SpotsRoute, s.handleSpots)
+		comms.RegisterHTTP(msgjson.CandlesRoute, s.handleCandles)
+		comms.RegisterHTTP(msgjson.OrderBookRoute, s.handleOrderBook)
+	}
+	return s
+}
+
+// AddMarketSource should be called before any markets are running.
+func (s *DataAPI) AddMarketSource(mkt MarketSource) error {
+	mktName, err := dex.MarketName(mkt.Base(), mkt.Quote())
+	if err != nil {
+		return err
+	}
+	epochDur := mkt.EpochDuration()
+	s.epochDurations[mktName] = epochDur
+	binCaches := make(map[uint64]*db.CandleCache, len(binSizes)+1)
+	s.marketCaches[mktName] = binCaches
+	cacheList := make([]*db.CandleCache, 0, len(binSizes)+1)
+	for _, binSize := range append([]uint64{epochDur}, binSizes...) {
+		cache := db.NewCandleCache(CacheSize, binSize)
+		cacheList = append(cacheList, cache)
+		binCaches[binSize] = cache
+		if binSize == bin5min {
+			s.cache5min = cache
+		}
+	}
+	if s.cache5min == nil {
+		panic("no 5-minute cache")
+	}
+	err = s.db.LoadEpochStats(mkt.Base(), mkt.Quote(), cacheList)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// SetBookSource should be called before the first call to handleBook.
+func (s *DataAPI) SetBookSource(bs BookSource) {
+	s.bookSource = bs
+}
+
+// ReportEpoch should be called by every Market after every match cycle to
+// report their epoch stats.
+func (s *DataAPI) ReportEpoch(base, quote uint32, epochIdx uint64, stats *matcher.MatchCycleStats) (*msgjson.Spot, error) {
```

Comment:
I've added the `Spot` return value here because I'd like clients to be able to subscribe to a feed of `Spots`. This would allow us to display little green arrows and percent changes next to the market names in the UI.